_posts/2017-11-21-javascript-modules.md

```diff
@@ -0,0 +1,216 @@
+
+In the JavaScript language, modules have been available natively only since 2015 (wow!) with the ECMAScript 6 [standard](http://www.ecma-international.org/ecma-262/6.0/ECMA-262.pdf). Not being able to break down code into modules is madness (in my opinion) and the community has come up with other ways to add modules to the language. (I am not really sure why modules were not added before, I don't know what the process is to create a new ECMAScript standard).
+<br/>
+The most famous module standards (pre ES6) are [AMD](https://github.com/amdjs/amdjs-api/blob/master/AMD.md)
+(a browser targeted standard whose most famous implementation is [requireJS](http://requirejs.org/)) and [CommonJS Modules](http://wiki.commonjs.org/wiki/Modules/1.1) (a standard defined by the group [CommonJS](https://en.wikipedia.org/wiki/CommonJS)).
+
+So depending on the context, people may refer to ES modules (ESM), CommonJS (CJS) modules or AMD (I don't think there is even a consistency in acronym 🤔).
+
+
+
+#### Using Modules in practice
+
+I think I have an understanding of what modules are or at least can refer to. What wasn't clear to me was how they were dealt with in my setup. I use [webpack 2](https://webpack.js.org/) and [Babel](https://babeljs.io/) and since modules are ES6 features I was wondering whether they will be converted to something (what??) by Babel or whether webpack will deal with them directly (but how??). I was super confused.
+
+So, let's have a look at a piece of code like below:
+
+```javascript
+// simple.js
+export let str = "Hello World";
+
+// main.js
+import { str } from './simple';
+console.log(str);
+```
+
+##### Babel
+
+To experiment with Babel I installed [babel-cli](https://babeljs.io/docs/usage/cli/) and configure it using a .babelrc config file. Then I run `npx babel-cli main.js simple.js -o output.js`.
+```json
+// .babelrc
+{
+    "presets": [
+        ["es2015"]
+    ]
+}
+```
+```javascript
+// output.js
+"use strict";
+
+Object.defineProperty(exports, "__esModule", {
+  value: true
+});
+var str = exports.str = "Hello World";
+'use strict';
+
+var _simple = require('./simple');
+
+console.log(_simple.str);
+```
+
+Ah that's very interesting! ES6 export statements are added as properties to an object called `exports` while ES6 import use a function called `require`. However, none of them is defined in the output file so where are they coming from?! And this is where knowing a bit about the history of JavaScript modules is useful. Babel's job is to transpile to ES5 and since ES5 has no native modules, it converts ES Modules to one of the pre ES6 modules standard (e.g. CommonJS, AMD). By default, it uses the CommonJS modules standard and if you check the standard [here](http://wiki.commonjs.org/wiki/Modules/1.1) you will see that `exports` and `require` are indeed defined!
```

Comment:
'none of them is' -> 'neither of them is'

Reply:
Neither of them _are_

Reply:
touché

Reply:
haha